Fix discovery when advertisement doesn't contain a useful name

jschornick opened the pull request:
Hi @ztroop! Thanks for creating a community version of the Airthings code.

When I gave your code a try with my 2nd generation Wave, sadly it didn't recognize the device. As far as I can tell, the problem is due to how you are relying on the BLE advertisement name to determine the model. 

With my device and on my system (Linux using BlueZ 5.55), the BLE name shows up as a fairly anonymous-looking variant on the MAC address (e.g. "12-34-56-78-90-AB"). Every once in a while, it would be populated with the real name "Airthings Wave2", but only reliably if I connected to the device first (such as via bluetoothctl).

Here's my attempt at a fix that reworks WaveProduct to be indexed by model number (instead of name), and modified detection code based on the decoded serial number. As far as I can tell from various photos online, the serial number always embeds the model as the first 4 digits.

I also added a test for the original issue, as well as a new data section for the 2nd generation Wave product.

jschornick committed:
Test BLE advertisements that lack "Wave*" name

jschornick committed:
Index WaveProducts by model number instead of name

jschornick committed:
Identify model from manufacturer data, not name

The BLE advertisement does not always contain useful data in the name
field. It may, for example, contain a variant of the MAC address
"18-04-ED-61-E4-4D", instead of a product name such as "Wave+".

To handle this situation more robustly, base the Airthings product
identification on the manufacturer data field. This should always be
present in a valid advertisement. For Wave devices, the manufacturer
data decodes to a serial number with the first 4 digits identifying the
specific model, e.g. "2930" for a Airthings Wave+.

ztroop commented on wave_reader/data.py:
```diff
@@ -26,6 +33,20 @@ class WaveProduct(Enum):
         ),
     },
     WaveProduct.WAVE: {
+        "name": "Wave",
```
Are we sure the `SENSOR_FORMAT` here will work for the First Gen Wave? It's not immediately obvious to me from looking at the Airthings [example code](https://github.com/Airthings/wave-reader/blob/master/read_wave.py).

ztroop replied:
The _First Gen_ Wave is something I was planning on adding "soon". Its `SENSOR_FORMAT` for `struct.unpack` would be `<H5B4H`, but haven't really dug into that yet. It's a challenge to support different devices without physically accessing the devices. :laughing: I think we're on the right track though.

jschornick replied:
Totally agree... perhaps the project will attract a few willing testers with access to the mystery devices. The Home Assistant community might be a good place to send out some feelers.

Back on topic, would it be kosher if I simply update the buffer format per your spec and return an empty dict for the time being? I'd rather not overcomplicate this PR by tackling multiple issues... especially ones we can't test just yet!

ztroop replied:
Great idea! And yes, that would be fine for the time being.

ztroop commented on wave_reader/data.py:
```diff
@@ -2,15 +2,22 @@
+
 DEVICE = {
     WaveProduct.WAVEPLUS: {
+        "name": "Wave+",
```
I think we should capitalise `name` _and below usage_, to be consistent with the rest of the key names.

jschornick replied:
No problem... I certainly should have followed the existing style. Though I have to ask, why *are* these keys all capitalized? I've never seen that style before.

ztroop replied:
I've always observed capitalisation in `Enum` members. I'm not sure if that's actually a convention, but seems to be consistent with [documentation](https://docs.python.org/3/library/enum.html) around `Enum`  usage. I don't feel strongly one way or the other, as long as it's consistent in the code.

jschornick replied:
Fair enough, but aren't `NAME` and friends the dict keys? Specifically, `WAVEPLUS` would be an enum member, sure... but `DEVICE` is simply a dict of dicts with a `WaveProduct` index.

But like I said, my bad for not following local convention. Fixed, and an updated commit is on its way.

ztroop commented on wave_reader/wave.py:
```diff
@@ -96,32 +96,36 @@ class WaveDevice:
+        self.model = self.serial[:4]
+        try:
+            self.product: WaveProduct = WaveProduct(self.model)
+        except KeyError:
```
If `self.model` does not match, it will throw `ValueError: 'self.name' is not a valid WaveProduct`. I don't think capturing `KeyError` is needed here.

jschornick replied:
D'oh! You're totally right. I think my brain was thinking dict and not enum. Happy to simplify the logic there. Nice catch.

ztroop reviewed (changes requested):
Thanks for the PR! :smile: Just a couple of things before merging.

jschornick committed:
Minor corrections after review of PR #2

Remove sensor parsing of original WAVE device for the time being, until
one is available for testing. Adjust test accordingly.

Eliminate useless exception handling around WaveProduct lookup.

More consistent formatting.
